BLD: Allow to set python path also via meson

doronbehar opened the pull request:
#### Reference issue

Helps gh-14812.

#### What does this implement/fix?

This helps cross compilation, as a user will want this path to be the host's python path.

#### Additional information

Nix cross compilation framework, didn't allow me keep this reference of the cross compiled scipy to the build platform's python executable.

rgommers commented:
Thanks @doronbehar. I'll have a closer look at this soon. First question: can you confirm the build/host terminology? The PR description says "build platform's" but you've named it `host-python`. The terminology we want to use is the one consistent with Meson:
- `build`: the machine/platform on which the build process is being executed
- `host`: the machine/platform on which the build artifacts are going to be installed and run

> This helps cross compilation, as a user will want this path to be the host's python path.

Second question: what do you actually do with this? You are introducing new a new `host_python_version` into `meson.build`, but it's not used for anything beyond some diagnostic output.

doronbehar commented:
> > Nix cross compilation framework, didn't allow me keep this reference of the cross compiled scipy to the build platform's python executable.
> The PR description says "build platform's" but you've named it `host-python`.

This was no mistake (pretty sure I got the terminology correct): The build platform's Python is the python you are building with - the Python I used to cross compile it with Nix, which is not the host python, that you want it's path to be put in that `config.py` file.

> > This helps cross compilation, as a user will want this path to be the host's python path.
> 
> Second question: what do you actually do with this? You are introducing a new `host_python_version` into `meson.build`, but it's not used for anything beyond some diagnostic output.

I'm personally don't do anything with this variable too :). But, if any Scipy developer found it reasonable to put this variable there in the first place, it might as well be correct when one cross compiles scipy. If you think perhaps it'd be better of removed alltogether, or only when one cross compiles scipy, that's also an option.

rgommers commented:
Thanks for the answers. I thought you needed this to successfully cross-compile, however that's not the case - you're only trying to improve the diagnostic output here. 

This PR is only a band-aid on the diagnostic output though, I think there's a problem here in the cross build support for Python in Meson, as discussed in https://github.com/mesonbuild/meson/issues/7049.

rgommers commented:
> This PR is only a band-aid on the diagnostic output though, I think there's a problem here in the cross build support for Python in Meson, as discussed in [mesonbuild/meson#7049](https://github.com/mesonbuild/meson/issues/7049).

I've converted this PR to draft, to indicate that it's blocked on a resolution to what the actual cross-compilation design for Python is in [mesonbuild/meson#7049](https://github.com/mesonbuild/meson/issues/7049).

doronbehar committed:
BLD: Allow to set python path also via meson

This helps cross compilation, as a user will want this path to be the
host's python path.

doronbehar commented:
Fixed merge conflicts with 1.18.1rc1 & branch `main`.

lucascolley commented:
@mgorny @rgommers am I right in thinking that we are waiting for some way forward via https://github.com/mesonbuild/meson-python/pull/829 ?

rgommers commented:
Yes, doing this via `build-details.json` sounds right. That'd be much better than having homegrown knobs specific to SciPy.
